pkg/ring/token_checker.go

```diff
@@ -0,0 +1,194 @@
+// ring and what its neighbors are.
+//
+// Checking that newly created streams fall within the expected token
+// ranges ensures that writes to the ingester are distributed properly,
+// while checking for validity of streams on an interval ensures that
+// an ingester's memory only contains appropriate tokens as the ingesters
+// in the ring change over time.
+type TokenChecker struct {
+	cfg        TokenCheckerConfig
+	lifecycler *Lifecycler
+
+	// Lifecycle control
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	// Updated throughout the lifetime of a TokenChecker
+	mut            sync.Mutex
+	expectedRanges []TokenRange
+
+	unexpectedStreamsHandler func(streamTokens []uint32)
+}
+
+// NewTokenChecker makes and starts a new TokenChecker. unexpectedStreamsHandler will be
+// invoked by TokenChecker whenever CheckAllTokens is called, even when no unexpected
+// tokens are found. If nil, unexpectedStreamsHandler is a no-op.
+func NewTokenChecker(cfg TokenCheckerConfig, lc *Lifecycler, unexpectedStreamsHandler func(streamTokens []uint32)) *TokenChecker {
+	tc := &TokenChecker{
+		cfg:                      cfg,
+		lifecycler:               lc,
+		unexpectedStreamsHandler: unexpectedStreamsHandler,
+	}
+	tc.ctx, tc.cancel = context.WithCancel(context.Background())
+
+	go tc.loop()
+	return tc
+}
+
+// Shutdown stops the Token Checker. It will stop watching the ring
+// for changes and stop checking that tokens are valid on an interval.
+func (tc *TokenChecker) Shutdown() {
+	tc.cancel()
+}
+
+// TokenExpected iterates over all expected ranges and returns true
+// when the token falls within one of those ranges.
+func (tc *TokenChecker) TokenExpected(token uint32) bool {
+	tc.mut.Lock()
+	defer tc.mut.Unlock()
+
+	for _, rg := range tc.expectedRanges {
+		if rg.Contains(token) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// CheckAllStreams invokes TokenExpected for all current untransferred
+// tokens found in the stored IncrementalTransferer. Invokes
+// tc.InvalidTokenHandler when an invalid token was found. Returns true when
+// all tokens are valid.
+func (tc *TokenChecker) CheckAllStreams() bool {
+	var invalid []uint32
+
+	toks := tc.lifecycler.incTransferer.MemoryStreamTokens()
+	for _, tok := range toks {
+		valid := tc.TokenExpected(tok)
+		if !valid {
+			invalid = append(invalid, tok)
+		}
+	}
+
+	numInvalid := len(invalid)
+
+	if tc.unexpectedStreamsHandler != nil {
+		tc.unexpectedStreamsHandler(invalid)
+	}
+
+	return numInvalid == 0
+}
+
+// syncRing syncs the latest ring from the lifecycler with expected ranges.
+func (tc *TokenChecker) syncRing() {
+	r := tc.lifecycler.getLastRing()
+	if r == nil {
+		// Wait for there to be a ring available to use
+		return
+	}
+	tc.updateExpectedRanges(r)
+}
+
+// updateExpectedRanges goes through the ring and finds all expected ranges
+// given the current set of tokens in a Lifecycler.
+func (tc *TokenChecker) updateExpectedRanges(ring *Desc) {
```

Comment:
Yes, it is OK. Even though C is in the ring, because it is unhealthy, A will be receiving all of C's writes from the distributor; A effectively owns C's ranges. When C is healthy again, this function will properly detect that and shorten the ranges for A.